src/main/java/dev/slimevr/vr/processor/skeleton/LegTweaks.java

```diff
@@ -870,6 +772,130 @@ private float calculateCorrectionWeight(
+	// returns a vector representing the center of mass position
+	private Vector3f computeCenterOfMass() {
+		// preform a check to see if the needed data is available
+		if (
+			skeleton.headNode == null
+				|| skeleton.chestNode == null
+				|| skeleton.waistNode == null
+				|| skeleton.leftFootNode == null
+				|| skeleton.rightFootNode == null
+				|| skeleton.leftKneeNode == null
+				|| skeleton.rightKneeNode == null
+				|| skeleton.leftHipNode == null
+				|| skeleton.rightHipNode == null
+
+		) {
+			return null;
+		}
+
+		// check if arm data is available
+		boolean armsAvaliable = true;
+		if (
+			!skeleton.hasLeftArmTracker
+				|| !skeleton.hasRightArmTracker
+		) {
+			armsAvaliable = false;
+		}
+
+		Vector3f centerOfMass = new Vector3f();
+
+		// compute the center of mass of smaller body parts and then sum them up
+		// with their respective weights
+		Vector3f head = skeleton.headNode.worldTransform.getTranslation();
+		Vector3f chest = skeleton.chestNode.worldTransform.getTranslation();
+		Vector3f waist = skeleton.waistNode.worldTransform.getTranslation();
+		Vector3f leftCalf = skeleton.leftAnkleNode.worldTransform
+			.getTranslation(null)
+			.add(skeleton.leftKneeNode.worldTransform.getTranslation(null))
+			.mult(0.5f);
+		Vector3f rightCalf = skeleton.rightAnkleNode.worldTransform
+			.getTranslation(null)
+			.add(skeleton.rightKneeNode.worldTransform.getTranslation(null))
+			.mult(0.5f);
+		Vector3f leftThigh = skeleton.leftKneeNode.worldTransform
+			.getTranslation(null)
+			.add(skeleton.leftHipNode.worldTransform.getTranslation(null))
+			.mult(0.5f);
+		Vector3f rightThigh = skeleton.rightKneeNode.worldTransform
+			.getTranslation(null)
+			.add(skeleton.rightHipNode.worldTransform.getTranslation(null))
+			.mult(0.5f);
+		centerOfMass = centerOfMass.add(head.mult(HEAD_MASS));
+		centerOfMass = centerOfMass.add(chest.mult(CHEST_MASS));
+		centerOfMass = centerOfMass.add(waist.mult(WAIST_MASS));
+		centerOfMass = centerOfMass.add(leftCalf.mult(CALF_MASS));
+		centerOfMass = centerOfMass.add(rightCalf.mult(CALF_MASS));
+		centerOfMass = centerOfMass.add(leftThigh.mult(THIGH_MASS));
+		centerOfMass = centerOfMass.add(rightThigh.mult(THIGH_MASS));
+		if (armsAvaliable) {
+			Vector3f leftUpperArm = skeleton.leftElbowNode.worldTransform
+				.getTranslation(null)
+				.add(skeleton.leftShoulderTailNode.worldTransform.getTranslation(null))
+				.mult(0.5f);
+			Vector3f rightUpperArm = skeleton.rightElbowNode.worldTransform
+				.getTranslation(null)
+				.add(skeleton.rightShoulderTailNode.worldTransform.getTranslation(null))
+				.mult(0.5f);
+			Vector3f leftForearm = skeleton.leftElbowNode.worldTransform
+				.getTranslation(null)
+				.add(skeleton.leftHandNode.worldTransform.getTranslation(null))
+				.mult(0.5f);
+			Vector3f rightForearm = skeleton.rightElbowNode.worldTransform
+				.getTranslation(null)
+				.add(skeleton.rightHandNode.worldTransform.getTranslation(null))
+				.mult(0.5f);
```

Comment:
maybe these calculations should be in a separate function, my copy paste detector throws warnings at me here